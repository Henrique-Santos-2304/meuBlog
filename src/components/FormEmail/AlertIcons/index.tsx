import React from "react";
import { Alert, AlertIcon } from "@chakra-ui/react";

type messageType = { message: string };
const AlertIcons = ({ message }: messageType) => {
  return (
    <Alert
      status="error"
      role="alert"
      colorScheme="purple"
      color="blackAlpha.900"
      borderRadius="1rem"
      mt="1rem"
    >
      <AlertIcon color="red.900" />
      {message}
    </Alert>
  );
};

export default AlertIcons;
