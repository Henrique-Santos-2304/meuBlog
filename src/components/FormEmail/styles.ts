import styled from "styled-components";
export const Form = styled.form`
  width: 70vw;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-bottom: 4rem;
`;
