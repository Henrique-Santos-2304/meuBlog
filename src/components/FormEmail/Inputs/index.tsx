/* eslint-disable @typescript-eslint/no-explicit-any */
import { FormControl, FormLabel, Input } from "@chakra-ui/react";
import { FieldValues, UseFormRegister } from "react-hook-form";
import AlertIcons from "../AlertIcons";

type IconsTypesProps = {
  txt: string;
  typeInput: string;
  registerer: UseFormRegister<FieldValues>;
  err: {
    [x: string]: any;
  };
};
const Inputs = ({ txt, typeInput, registerer, err }: IconsTypesProps) => {
  const emailPattern =
    typeInput === "email"
      ? {
          required: "Campo Obrigatório",
          pattern: {
            value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
            message: "invalid email address",
          },
        }
      : { required: "Campo Obrigatório" };
  return (
    <FormControl>
      <FormLabel
        fontSize="2xl"
        htmlFor={typeInput}
        color="yellow.600"
        ml="0.5rem"
      >
        {txt}
      </FormLabel>
      <Input
        isRequired={false}
        variant="flushed"
        textAlign="center"
        type="text"
        placeholder={`Digite seu ${txt}`}
        _focus={{ borderColor: "orange" }}
        {...registerer(typeInput, emailPattern)}
        fontSize="2xl"
        borderRadius="1rem"
        color="gray.100"
        id={typeInput}
        h="2.6rem"
      />
      {err[typeInput] && <AlertIcons message={err[typeInput].message} />}
    </FormControl>
  );
};

export default Inputs;
