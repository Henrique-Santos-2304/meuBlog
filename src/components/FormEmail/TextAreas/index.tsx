/* eslint-disable @typescript-eslint/no-explicit-any */
import { FormControl, FormLabel, Textarea } from "@chakra-ui/react";
import { FieldValues, UseFormRegister } from "react-hook-form";
import AlertIcons from "../AlertIcons";

type IconsTypesProps = {
  txt: string;
  typeInput: string;
  registerer: UseFormRegister<FieldValues>;
  err: {
    [x: string]: any;
  };
};
const TextAreas = ({ txt, typeInput, registerer, err }: IconsTypesProps) => {
  return (
    <FormControl>
      <FormLabel fontSize="2xl" htmlFor={typeInput} color="yellow.600">
        {txt}
      </FormLabel>
      <Textarea
        {...registerer(typeInput, {
          required: `Campo  Obrigatório`,
        })}
        placeholder={`Digite sua Mensagem`}
        name={typeInput}
        aria-label="Campo de Digitação de Mensagem"
        type={typeInput}
        h="17rem"
        p="1.3rem"
        variant="filed"
        bg="blackAlpha.300"
        color="gray.100"
        textAlign="center"
        fontSize="2xl"
        borderRadius="1.5rem"
        transition="all .3s linear"
        _focus={{ border: "0.1rem solid orange" }}
        _hover={{ border: "0.1rem solid orange" }}
      />
      {err[typeInput] && <AlertIcons message={err[typeInput].message} />}
    </FormControl>
  );
};

export default TextAreas;
