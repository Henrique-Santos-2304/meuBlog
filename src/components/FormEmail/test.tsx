import FormEmail from ".";
import { screen, fireEvent } from "@testing-library/react";
import { renderTheme } from "utils/testRenderTheme";
import { mocked } from "ts-jest/utils";
import emailjs from "emailjs-com";

jest.mock("emailjs-com");

describe("<FormEmail />", () => {
  it("should render ", () => {
    const { container } = renderTheme(<FormEmail />);
    expect(container).toMatchSnapshot();
  });

  it("all fields of the imaginary form on the screen", () => {
    renderTheme(<FormEmail />);
    expect(screen.getByText(/nome/i)).toBeInTheDocument();
    expect(screen.getByText(/email/i)).toBeInTheDocument();
    expect(screen.getByText(/mensagem/i)).toBeInTheDocument();
    expect(
      screen.getByRole("textbox", {
        name: /nome/i,
      })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("textbox", {
        name: /email/i,
      })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("textbox", {
        name: /campo de digitação de mensagem/i,
      })
    ).toBeInTheDocument();
  });

  it("should Have 3 alerts with 3 invalid fields", async () => {
    renderTheme(<FormEmail />);
    const button = screen.getByRole("button", { name: /enviar/i });
    expect(button).toBeInTheDocument();

    fireEvent.submit(button);
    expect(await screen.findAllByRole("alert")).toHaveLength(3);
  });

  it("should have invalid name alert", async () => {
    renderTheme(<FormEmail />);
    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@test.com",
      },
    });
    fireEvent.input(screen.getByRole("textbox", { name: /nome/i }), {
      target: {
        value: "",
      },
    });
    fireEvent.input(
      screen.getByRole("textbox", { name: /campo de digitação de mensagem/i }),
      {
        target: {
          value: "asdasd",
        },
      }
    );
    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
  });

  it("should have invalid email alert", async () => {
    renderTheme(<FormEmail />);
    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test",
      },
    });
    fireEvent.input(screen.getByRole("textbox", { name: /nome/i }), {
      target: {
        value: "nome",
      },
    });
    fireEvent.input(
      screen.getByRole("textbox", { name: /campo de digitação de mensagem/i }),
      {
        target: {
          value: "message",
        },
      }
    );
    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
  });

  it("should have invalid message alert", async () => {
    renderTheme(<FormEmail />);
    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@test.com",
      },
    });
    fireEvent.input(screen.getByRole("textbox", { name: /nome/i }), {
      target: {
        value: "asdasdd",
      },
    });
    fireEvent.input(
      screen.getByRole("textbox", { name: /campo de digitação de mensagem/i }),
      {
        target: {
          value: "",
        },
      }
    );
    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
  });

  it("should have sent message Sucess when the data is valid", async () => {
    const mockEmaijs = mocked(emailjs);

    renderTheme(<FormEmail />);
    fireEvent.input(screen.getByRole("textbox", { name: /nome/i }), {
      target: {
        value: "test",
      },
    });
    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(
      screen.getByRole("textbox", { name: /campo de digitação de mensagem/i }),
      {
        target: {
          value: "Mensagem",
        },
      }
    );

    fireEvent.submit(screen.getByRole("button", { name: /enviar/i }));
    expect(await screen.queryAllByRole("alert")).toHaveLength(0);
    await mockEmaijs.sendForm.mockResolvedValueOnce({
      text: "OK",
      status: 200,
    });

    expect(await screen.findByText(/Enviado com sucesso/i)).toBeInTheDocument();
  });
  it("should have sent message error when the data is invalid", async () => {
    const mockEmaijs = mocked(emailjs);

    renderTheme(<FormEmail />);
    fireEvent.input(screen.getByRole("textbox", { name: /nome/i }), {
      target: {
        value: "test",
      },
    });
    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(
      screen.getByRole("textbox", { name: /campo de digitação de mensagem/i }),
      {
        target: {
          value: "Mensagem",
        },
      }
    );

    fireEvent.submit(screen.getByRole("button", { name: /enviar/i }));
    expect(await screen.queryAllByRole("alert")).toHaveLength(1);
    mockEmaijs.sendForm.mockRejectedValueOnce({ Error });

    expect(
      await screen.findByText(/Erro!Por favor verifique seus dados/i)
    ).toBeInTheDocument();
  });
});
