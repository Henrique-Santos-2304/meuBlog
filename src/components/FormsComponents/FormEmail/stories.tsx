import { Story, Meta } from "@storybook/react";
import FormEmail from ".";

export default {
  title: "FormEmail",
  component: FormEmail,
} as Meta;

export const Basic: Story = (args) => <FormEmail {...args} />;

Basic.args = {
  title: "title basic",
  description: "description basic",
};

export const Default: Story = (args) => <FormEmail {...args} />;
