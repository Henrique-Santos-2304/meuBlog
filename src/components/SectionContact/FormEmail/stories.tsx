import { Story, Meta } from "@storybook/react";
import FormEmail from ".";
import { mockMainContact } from "../MainContact/mockMainContact";

export default {
  title: "FormEmail",
  component: FormEmail,
} as Meta;

export const Basic: Story = () => <FormEmail {...mockMainContact} />;

export const Default: Story = () => <FormEmail {...mockMainContact} />;
