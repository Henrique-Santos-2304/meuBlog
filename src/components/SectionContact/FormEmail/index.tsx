/* eslint-disable @typescript-eslint/no-non-null-assertion */
import * as S from "./styles";
import { useEffect, useRef, useState } from "react";
import { useForm } from "react-hook-form";
import emailjs from "emailjs-com";
import ContentInput from "../ContentInput";

const FormEmail = () => {
  const {
    handleSubmit,
    register,
    formState: { errors },
  } = useForm();
  const formRef = useRef<HTMLFormElement>(null);
  const [msgForm, setMsgForm] = useState("");

  const onSubmit = async () => {
    await emailjs
      .sendForm(
        "service_ijjwv7k",
        "CONTACT_FORM",
        formRef.current!,
        "user_mB0pVxnJ1ebEVquIzWsUn"
      )
      .then(
        function () {
          setMsgForm("Enviado com sucesso");
        },
        function () {
          setMsgForm("Erro! Verifique seus dados");
        }
      );
  };
  useEffect(() => {
    setTimeout(() => {
      setMsgForm("");
    }, 10000);
  }, [msgForm]);

  return (
    <S.Wrapper>
      {msgForm && (
        <S.ModalForm aria-label="alerta de sucesso ou erro para envio de email">
          {msgForm}
        </S.ModalForm>
      )}
      <S.Form onSubmit={handleSubmit(onSubmit)} ref={formRef}>
        {/* Input de Nome */}
        <ContentInput
          error={errors.name && errors.name.message}
          typeLabel="name"
          txtLabel="Nome"
        >
          <S.Inputs
            id="name"
            type="text"
            placeholder="Digite seu nome"
            {...register("name", { required: "Digite seu nome" })}
          />
        </ContentInput>
        {/* Input de Email */}
        <ContentInput
          error={errors.email && errors.email.message}
          typeLabel="email"
          txtLabel="Email"
        >
          <S.Inputs
            id="email"
            type="email"
            placeholder="Digite seu email"
            {...register("email", {
              required: "Digite um email",
              pattern: {
                value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
                message: "Formato de email invÃ¡lido",
              },
            })}
          />
        </ContentInput>

        {/* Input de Mensagem  */}
        <ContentInput
          error={errors.message && errors.message.message}
          typeLabel="message"
          txtLabel="Mensagem"
        >
          <S.TextAreas
            id="message"
            placeholder="Digite sua mensagem"
            {...register("message", { required: "Digite uma mensagem" })}
          />
        </ContentInput>

        <S.Button type="submit">Enviar</S.Button>
      </S.Form>
    </S.Wrapper>
  );
};

export default FormEmail;
