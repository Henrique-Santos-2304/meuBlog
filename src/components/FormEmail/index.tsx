/* eslint-disable prettier/prettier */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import * as S from "./styles";
import { useForm } from "react-hook-form";
import { ArrowForwardIcon } from "@chakra-ui/icons";
import { Button, Flex, VStack, useToast } from "@chakra-ui/react";
import Inputs from "./Inputs";
import TextAreas from "./TextAreas";

type dataMesssage = {
  name: string;
  email: string;
  message: string;
};
const FormEmail = () => {
  const { handleSubmit, register, formState } = useForm();

  const toast = useToast();

  const onSubmit = (data: dataMesssage) => {
    toast({
      title: "Enviado!",
      status: "success",
      duration: 3000,
      isClosable: true,
    });
    console.log(data);
  };
  return (
    <Flex direction="column" align="center" justify="center">
      <S.Form onSubmit={handleSubmit(onSubmit)}>
        <VStack spacing={12} w="100%">
          {/* Campo de Nome */}
          <Inputs
            txt="Nome"
            type="name"
            registerer={register}
            err={formState.errors}
          />

          <Inputs
            txt="Email"
            type="email"
            registerer={register}
            err={formState.errors}
          />
          {/* Campo de Mensagem  */}
          <TextAreas
            txt="Mensagem"
            type="message"
            registerer={register}
            err={formState.errors}
          />

          <Button
            type="submit"
            rightIcon={<ArrowForwardIcon />}
            colorScheme="yellow"
            borderRadius="1rem"
            variant="outline"
            fontSize="2xl"
            size="lg"
            w="13rem"
            h="4.8rem"
            mt={4}
            _hover={{ filter: "brightness(0.8)" }}
            isLoading={formState.isSubmitting}
          >
            Enviar
          </Button>
        </VStack>
      </S.Form>
    </Flex>
  );
};

export default FormEmail;
// pattern: {
//   value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
//   message: "invalid email address",
// },
