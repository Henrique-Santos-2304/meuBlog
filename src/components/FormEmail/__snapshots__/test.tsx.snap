// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormEmail /> should render  1`] = `
.c0 {
  width: 70vw;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-bottom: 4rem;
}

<div>
   
  <div
    class="css-u6v8er"
  >
    <form
      class="c0"
    >
      <div
        class="chakra-stack css-r5xrk8"
      >
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-bk1051"
            for="name"
            id="field-1-label"
          >
            Nome
          </label>
          <input
            class="chakra-input css-14xgnea"
            id="name"
            name="name"
            placeholder="Digite seu Nome"
            type="name"
          />
        </div>
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-bk1051"
            for="email"
            id="field-2-label"
          >
            Email
          </label>
          <input
            class="chakra-input css-14xgnea"
            id="email"
            name="email"
            placeholder="Digite seu Email"
            type="email"
          />
        </div>
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-bv8qsa"
            for="message"
            id="field-3-label"
          >
            Mensagem
          </label>
          <textarea
            aria-label="Campo de Digitação de Mensagem"
            class="chakra-textarea css-1r5n2ve"
            id="field-3"
            name="message"
            placeholder="Digite sua Mensagem"
            type="message"
          />
        </div>
        <button
          class="chakra-button css-1i76mv4"
          type="submit"
        >
          Enviar
          <span
            class="chakra-button__icon css-1hzyiq5"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-onkibi"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                fill="currentColor"
              />
            </svg>
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
